api/ocm/extensions/repositories/genericocireg/accessmethod_localblob.go

```diff
@@ -111,3 +124,107 @@ func (m *localBlobAccessMethod) Get() ([]byte, error) {
 func (m *localBlobAccessMethod) MimeType() string {
 	return m.spec.MediaType
 }
+
+////////////////////////////////////////////////////////////////////////////////
+
+type composedBlock struct {
+	m    *localBlobAccessMethod
+	refs []string
+}
+
+var _ blobaccess.DataAccess = (*composedBlock)(nil)
+
+func (c *composedBlock) Get() ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	for _, ref := range c.refs {
+		var finalize finalizer.Finalizer
+
+		_, data, err := c.m.namespace.GetBlobData(digest.Digest(ref))
+		if err != nil {
+			return nil, err
+		}
+		finalize.Close(data)
+		r, err := data.Reader()
+		if err != nil {
+			return nil, err
+		}
+		finalize.Close(r)
+		_, err = io.Copy(buf, r)
+		if err != nil {
+			return nil, err
+		}
+		err = finalize.Finalize()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+func (c *composedBlock) Reader() (io.ReadCloser, error) {
+	return &composedReader{
+		m:    c.m,
+		refs: c.refs,
+	}, nil
+}
+
+func (c composedBlock) Close() error {
+	return nil
+}
+
+type composedReader struct {
+	lock   sync.Mutex
+	m      *localBlobAccessMethod
+	refs   []string
+	reader io.ReadCloser
+	data   blobaccess.DataAccess
+}
+
+func (c *composedReader) Read(p []byte) (n int, err error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for {
+		if c.reader != nil {
+			n, err := c.reader.Read(p)
+			if n > 0 {
+				if err == io.EOF {
+					err = nil
+				}
+				return n, err
+			}
+			if err != nil {
+				return n, err
+			}
+			c.reader.Close()
+			c.data.Close()
+			c.reader = nil
+		}
+		if len(c.refs) == 0 {
+			return 0, io.EOF
+		}
+
+		ref := strings.TrimSpace(c.refs[0])
+		_, c.data, err = c.m.namespace.GetBlobData(digest.Digest(ref))
+		if err != nil {
+			return 0, err
+		}
+		c.reader, err = c.data.Reader()
+		if err != nil {
+			return 0, err
+		}
+	}
```

Comment:
This loop is weird. I'm not sure what the intention is. It never removes the `refs` so while there is an empty reader, this is potentially an endless loop.

Handling of EOF only happens if there was a read `if n > 0` you should check it before that in an error case.

You are using p to read but you have no idea if there is capacity for it. Why?

So:
- check eof in case of error
- remove refs
- reader close only happens when there was a read but not if it's eof or if there was an error.

Reply:
Besides, `p` is not large enough to store the entire data, the reader is closed afterwards and I cannot read the rest of it or am I interpreting the code wrong?

Reply:
you are right, refs is never shrinked, strange, why are the tests succeeding

Reply:
ahh, another method, streaming is not tested :-(

Reply:
The final reader close is potentially done with the external `Close` of the `composedReader`

Reply:
no data get, streamed read and steamed read with small buffer is tested.

Reply:
nice, thanks